pkl-core/src/main/java/org/pkl/core/IniRenderer.java

```diff
@@ -0,0 +1,247 @@
+/**
+ * Copyright © 2024 Apple Inc. and the Pkl project authors. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.pkl.core;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.io.Writer;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.regex.Pattern;
+import org.pkl.core.util.Nullable;
+import org.pkl.core.util.ini.IniUtils;
+
+// To instantiate this class, use ValueRenderers.properties().
+final class IniRenderer implements ValueRenderer {
+
+  private final Writer writer;
+  private final boolean omitNullProperties;
+  private final boolean restrictCharset;
+
+  public IniRenderer(Writer writer, boolean omitNullProperties, boolean restrictCharset) {
+    this.writer = writer;
+    this.omitNullProperties = omitNullProperties;
+    this.restrictCharset = restrictCharset;
+  }
+
+  @Override
+  public void renderDocument(Object value) {
+    new Visitor().renderDocument(value);
+  }
+
+  @Override
+  public void renderValue(Object value) {
+    new Visitor().renderValue(value);
+  }
+
+  public class Visitor implements ValueConverter<String> {
+
+    public void renderValue(Object value) {
+      write(convert(value), false, restrictCharset);
+    }
+
+    public void renderDocument(Object value) {
+      if (value instanceof Composite) {
+        doVisitMap(null, ((Composite) value).getProperties());
+      } else if (value instanceof Map) {
+        doVisitMap(null, (Map<?, ?>) value);
+      } else if (value instanceof Pair) {
+        Pair<?, ?> pair = (Pair<?, ?>) value;
+        doVisitKeyAndValue(null, pair.getFirst(), pair.getSecond());
+      } else {
```

Comment:
To be consistent with other renderers, let's not support rendering pairs.

```suggestion
      } else {
```